### ISSUE TITLE
[Event Request] Report 99001025 "Refresh Production Order"

### ISSUE THREAD

**Issue by nevenamark** [microsoft/ALAppExtensions#27110]
### Describe the request

Hello. We need two events in trigger OnAfterGetRecord() and one in trigger OnPreDataItem() in report 99001025 "Refresh Production Order" respectively as follows:

```
[IntegrationEvent(false, false)]
local procedure OnBeforeConfirmPickedComponents(var IsHandled: Boolean)
begin
end;

[IntegrationEvent(false, false)]
local procedure OnBeforeUpdateWindow(var IsHandled: Boolean)
begin
end;

[IntegrationEvent(false, false)]
local procedure OnBeforeOpenWindow(var IsHandled: Boolean)
begin
end;

```
After implementing these changes relevant triggers would look like this:

```
trigger OnAfterGetRecord()
            var
                ProdOrderLine: Record "Prod. Order Line";
                ProdOrderRtngLine: Record "Prod. Order Routing Line";
                ProdOrderComp: Record "Prod. Order Component";
                ProdOrder: Record "Production Order";
                ProdOrderStatusMgt: Codeunit "Prod. Order Status Management";
                RoutingNo: Code[20];
                ErrorOccured: Boolean;
                IsHandled: Boolean;
            begin
                if Status = Status::Finished then
                    CurrReport.Skip();

                TestField("Due Date");

                if CalcLines and IsComponentPicked("Production Order") then begin
                    OnBeforeConfirmPickedComponents(IsHandled);
                    if not IsHandled then
                       if not Confirm(StrSubstNo(DeletePickedLinesQst, "No.")) then
                           CurrReport.Skip();
                end;

                OnBeforeUpdateWindow(IsHandled);
                if not IsHandled then begin
                   Window.Update(1, Status);
                   Window.Update(2, "No.");
                end;

                RoutingNo := GetRoutingNo("Production Order");
                UpdateRoutingNo("Production Order", RoutingNo);

                ProdOrderLine.LockTable();
                OnBeforeCalcProdOrder("Production Order", Direction);
                CheckReservationExist();

                if CalcLines then begin
                    OnBeforeCalcProdOrderLines("Production Order", Direction, CalcLines, CalcRoutings, CalcComponents, IsHandled, ErrorOccured);
                    if not IsHandled then
                        if not CreateProdOrderLines.Copy("Production Order", Direction, "Production Order"."Variant Code", false) then
                            ErrorOccured := true;
                end else begin
                    ProdOrderLine.SetRange(Status, Status);
                    ProdOrderLine.SetRange("Prod. Order No.", "No.");
                    IsHandled := false;
                    OnBeforeCalcRoutingsOrComponents("Production Order", ProdOrderLine, CalcComponents, CalcRoutings, IsHandled);
                    if not IsHandled then
                        if CalcRoutings or CalcComponents then begin
                            if ProdOrderLine.Find('-') then
                                repeat
                                    if CalcRoutings then begin
                                        ProdOrderRtngLine.SetRange(Status, Status);
                                        ProdOrderRtngLine.SetRange("Prod. Order No.", "No.");
                                        ProdOrderRtngLine.SetRange("Routing Reference No.", ProdOrderLine."Routing Reference No.");
                                        ProdOrderRtngLine.SetRange("Routing No.", ProdOrderLine."Routing No.");
                                        if ProdOrderRtngLine.FindSet(true) then
                                            repeat
                                                ProdOrderRtngLine.SetSkipUpdateOfCompBinCodes(true);
                                                ProdOrderRtngLine.Delete(true);
                                            until ProdOrderRtngLine.Next() = 0;
                                    end;
                                    if CalcComponents then begin
                                        ProdOrderComp.SetRange(Status, Status);
                                        ProdOrderComp.SetRange("Prod. Order No.", "No.");
                                        ProdOrderComp.SetRange("Prod. Order Line No.", ProdOrderLine."Line No.");
                                        ProdOrderComp.DeleteAll(true);
                                    end;
                                until ProdOrderLine.Next() = 0;
                            if ProdOrderLine.Find('-') then
                                repeat
                                    if CalcComponents then
                                        CheckProductionBOMStatus(ProdOrderLine."Production BOM No.", ProdOrderLine."Production BOM Version Code");
                                    if CalcRoutings then
                                        CheckRoutingStatus(ProdOrderLine."Routing No.", ProdOrderLine."Routing Version Code");
                                    ProdOrderLine."Due Date" := "Due Date";
                                    IsHandled := false;
                                    OnBeforeCalcProdOrderLine(ProdOrderLine, Direction, CalcLines, CalcRoutings, CalcComponents, IsHandled, ErrorOccured);
                                    if not IsHandled then
                                        if not CalcProdOrder.Calculate(ProdOrderLine, Direction, CalcRoutings, CalcComponents, false, false) then
                                            ErrorOccured := true;
                                until ProdOrderLine.Next() = 0;
                        end;
                end;
                OnProductionOrderOnAfterGetRecordOnAfterCalcRoutingsOrComponents("Production Order", CalcLines, CalcRoutings, CalcComponents, ErrorOccured);

                if (Direction = Direction::Backward) and ("Source Type" = "Source Type"::Family) then begin
                    SetUpdateEndDate();
                    Validate("Due Date", "Due Date");
                end;

                if Status = Status::Released then begin
                    ProdOrderStatusMgt.FlushProdOrder("Production Order", Status, WorkDate());
                    WhseProdRelease.Release("Production Order");
                    if CreateInbRqst then
                        WhseOutputProdRelease.Release("Production Order");
                end;

                OnAfterRefreshProdOrder("Production Order", ErrorOccured);
                if ErrorOccured then
                    Message(Text005, ProdOrder.TableCaption(), ProdOrderLine.FieldCaption("Bin Code"));
            end;

  trigger OnPreDataItem()
  var
     IsHandled: Boolean;
  begin
     OnBeforeOpenWindow(IsHandled);
     if not IsHandled then
                Window.Open(
                  Text000 +
                  Text001 +
                  Text002);
  end;
```

### Additional context

We need to add the condition in order to control whether the dialog and window get displayed. Adding requested event in the PreDataItem() trigger would imply a local variable IsHandled to create the condition as well.
Internal work item: [AB#545257](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/545257)

**Comment by JakovljevicDusan** [comment]
Implemented a bit different solution with HideValidationDialog "pattern"

**Comment by nevenamark** [comment]
For a different proposal of adding the HideValidationDialog variable and its setter procedure to the standard report, should I create a separate request?

Thank you in advance!

**Comment by JakovljevicDusan** [comment]
Hi @nevenamark ,

At the moment, I didn't want to refactor base app too much (then it will be shipped in 26.0), so this will be shipped in 25.1:
1. There is a public procedure SetHideValidationDialog, but it is not really used in Base app.
2. You can also set HideValidationDialog as new parameter in publisher OnAfterOnInit. You will probably use this.
3. Since when HideValidationDialog = true, there is autoconfirm logic, there is new event publisher to alter it - OnProductionOrderOnAfterGetRecordOnBeforeAlreadyPickedLinesConfirm (this is pretty much the same as OnBeforeConfirmPickedComponents you requested).

The same changes will be implemented in next major bracnsh, so you can also check in Base app main branch copy in GitHub in a view days.
There will be some other minor refactoring, like variable renaming, but not related to this request or execution logic.

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 25201.
